<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Trading Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/DOGEBUSD_trades_1m.csv" target="_blank">https://1824251045.rsc.cdn77.org/web/algohouse/data/DOGEBUSD_trades_1m.csv</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


you are running the lab in your  browser, so we will install the libraries using `piplite`


In [116]:
#you are running the lab in your  browser, so we will install the libraries using ``piplite``
import piplite
import micropip
await piplite.install(['pandas'])
await piplite.install(['matplotlib'])
await piplite.install(['scipy'])
await piplite.install(['seaborn'])
await micropip.install(['ipywidgets'],keep_going=True)
await micropip.install(['tqdm'],keep_going=True)

If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [117]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [118]:
# import pandas library
import pandas as pd
import numpy as np

#Further specify the value of the precision parameter equal to 2 to display two decimal signs (instead of 6 as default).
pd.set_option("display.precision", 2)
pd.options.display.float_format = '{:.2f}'.format

This function will download the dataset into your browser


In [119]:
#This function will download the dataset into your browser 

from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

You can also assign the dataset to any variable you create.

</p>


In [120]:
path = "https://1824251045.rsc.cdn77.org/web/algohouse/data/DOGEBUSD_trades_1m.csv"

you will need to download the dataset; if you are running locally, please comment out the following


In [121]:
#you will need to download the dataset; if you are running locally, please comment out the following 
await download(path, "DOGEBUSD_trades_1m.csv")
path="DOGEBUSD_trades_1m.csv"

This dataset was hosted on IBM Cloud object. Click <a href="https://cocl.us/DA101EN_object_storage?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01">HERE</a> for free storage.


In [122]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"
df = pd.read_csv(path, index_col=0)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [123]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


ts  open  high  low  close     volume  rec_count  \
0  2022-11-11 14:38:00  0.08  0.08 0.08   0.08 4523754.00         80   
1  2022-11-11 14:39:00  0.08  0.08 0.08   0.08 5119775.00         77   
2  2022-11-11 14:40:00  0.08  0.08 0.08   0.08 1460872.00         65   
3  2022-11-11 14:41:00  0.08  0.08 0.08   0.08 2292152.00         75   
4  2022-11-11 14:42:00  0.08  0.08 0.08   0.08 3044077.00         57   

   avg_price  
0       0.08  
1       0.08  
2       0.08  
3       0.08  
4       0.08

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [124]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Add Headers</h3>
<p>
Take a look at our dataset. Pandas automatically set the header with an integer starting from 0.
</p>

To better describe our data, we can introduce a header.
</p>

<p>
Thus, we have to add headers manually.
</p>

<p>
First, we create a list "headers" that include all column names in order.
Then, we use <code>dataframe.columns = headers</code> to replace the headers with the list we created.
</p>


In [135]:
# create headers list
headers = ["Ts", "Open", "High", "Low", "Close", "Volume", "Rec_count", "Avg_price"]
print("headers\n", headers)

headers
 ['Ts', 'Open', 'High', 'Low', 'Close', 'Volume', 'Rec_count', 'Avg_price']


We replace headers and recheck our dataframe:


In [136]:
df.columns = headers
df.head(10)

Ts  Open  High  Low  Close      Volume  Rec_count  \
0  2022-11-11 14:38:00  0.08  0.08 0.08   0.08  4523754.00         80   
1  2022-11-11 14:39:00  0.08  0.08 0.08   0.08  5119775.00         77   
2  2022-11-11 14:40:00  0.08  0.08 0.08   0.08  1460872.00         65   
3  2022-11-11 14:41:00  0.08  0.08 0.08   0.08  2292152.00         75   
4  2022-11-11 14:42:00  0.08  0.08 0.08   0.08  3044077.00         57   
5  2022-11-11 14:43:00  0.08  0.08 0.08   0.08  1476278.00         66   
6  2022-11-11 14:44:00  0.08  0.08 0.08   0.08  1838426.00         60   
7  2022-11-11 14:45:00  0.08  0.08 0.08   0.08 13747027.00        100   
8  2022-11-11 14:46:00  0.08  0.08 0.08   0.08 12758340.00         93   
9  2022-11-11 14:47:00  0.08  0.09 0.08   0.08 34452376.00        105   

   Avg_price  
0       0.08  
1       0.08  
2       0.08  
3       0.08  
4       0.08  
5       0.08  
6       0.08  
7       0.08  
8       0.08  
9       0.08

We can drop missing values along the column "open" by uncommenting the following:


In [128]:
#df=df1.dropna(subset=["price"], axis=0)
df.head(20)

ts  open  high  low  close      volume  rec_count  \
0   2022-11-11 14:38:00  0.08  0.08 0.08   0.08  4523754.00         80   
1   2022-11-11 14:39:00  0.08  0.08 0.08   0.08  5119775.00         77   
2   2022-11-11 14:40:00  0.08  0.08 0.08   0.08  1460872.00         65   
3   2022-11-11 14:41:00  0.08  0.08 0.08   0.08  2292152.00         75   
4   2022-11-11 14:42:00  0.08  0.08 0.08   0.08  3044077.00         57   
5   2022-11-11 14:43:00  0.08  0.08 0.08   0.08  1476278.00         66   
6   2022-11-11 14:44:00  0.08  0.08 0.08   0.08  1838426.00         60   
7   2022-11-11 14:45:00  0.08  0.08 0.08   0.08 13747027.00        100   
8   2022-11-11 14:46:00  0.08  0.08 0.08   0.08 12758340.00         93   
9   2022-11-11 14:47:00  0.08  0.09 0.08   0.08 34452376.00        105   
10  2022-11-11 14:48:00  0.08  0.09 0.08   0.08 25284894.00        106   
11  2022-11-11 14:49:00  0.08  0.08 0.08   0.08 18461593.00        103   
12  2022-11-11 14:50:00  0.08  0.08 0.08   0.08  6608245.00         99   
13  2022-11-11 14:51:00  0.08  0.08 0.08   0.08  9624168.00         97   
14  2022-11-11 14:52:00  0.08  0.08 0.08   0.08 14723866.00        102   
15  2022-11-11 14:53:00  0.08  0.08 0.08   0.08  6915675.00         98   
16  2022-11-11 14:54:00  0.08  0.08 0.08   0.08  3355101.00         79   
17  2022-11-11 14:55:00  0.08  0.08 0.08   0.08  3792184.00         85   
18  2022-11-11 14:56:00  0.08  0.08 0.08   0.08  8703982.00         89   
19  2022-11-11 14:57:00  0.08  0.08 0.08   0.08  3535313.00         82   

    avg_price  
0        0.08  
1        0.08  
2        0.08  
3        0.08  
4        0.08  
5        0.08  
6        0.08  
7        0.08  
8        0.08  
9        0.08  
10       0.08  
11       0.08  
12       0.08  
13       0.08  
14       0.08  
15       0.08  
16       0.08  
17       0.08  
18       0.08  
19       0.08

Now, we have successfully read the raw dataset and added the correct headers into the dataframe.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [129]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>DOGEBUSD_trades_1m.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

</p>


In [137]:
df.dtypes

Ts            object
Open         float64
High         float64
Low          float64
Close        float64
Volume       float64
Rec_count      int64
Avg_price    float64
dtype: object

<p>
As shown above, it is clear to see that the data type of "Open" and "Close" are <code>float64</code>, "Ts" is <code>object</code>, and "Rec_count" is <code>int64</code>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [138]:
df.describe()

Open     High      Low    Close      Volume  Rec_count  Avg_price
count 67277.00 67277.00 67277.00 67277.00    67277.00   67277.00   67277.00
mean      0.09     0.09     0.09     0.09  1329309.25      26.86       0.09
std       0.01     0.01     0.01     0.01  2587028.18      18.22       0.01
min       0.07     0.07     0.07     0.07        8.00       1.00       0.07
25%       0.08     0.08     0.08     0.08   277885.00      13.00       0.08
50%       0.09     0.09     0.09     0.09   636292.00      22.00       0.09
75%       0.10     0.10     0.10     0.10  1402308.00      36.00       0.10
max       0.11     0.11     0.11     0.11 74130181.00     119.00       0.11

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "Open" has 67277 counts, the mean value of this column is 0.09, the standard deviation is 0.01, the minimum value is 0.07, 25th percentile is 0.08, 50th percentile is 0.09, 75th percentile is 0.10, and the maximum value is 0.11. <br>

However, what if we would also like to check all the columns including those that are of type object? <br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [132]:
# describe all the columns in "df" 
df.describe(include = "all", datetime_is_numeric = True)

ts     open     high      low    close      volume  \
count                 67277 67277.00 67277.00 67277.00 67277.00    67277.00   
unique                67277      NaN      NaN      NaN      NaN         NaN   
top     2022-11-11 14:38:00      NaN      NaN      NaN      NaN         NaN   
freq                      1      NaN      NaN      NaN      NaN         NaN   
mean                    NaN     0.09     0.09     0.09     0.09  1329309.25   
std                     NaN     0.01     0.01     0.01     0.01  2587028.18   
min                     NaN     0.07     0.07     0.07     0.07        8.00   
25%                     NaN     0.08     0.08     0.08     0.08   277885.00   
50%                     NaN     0.09     0.09     0.09     0.09   636292.00   
75%                     NaN     0.10     0.10     0.10     0.10  1402308.00   
max                     NaN     0.11     0.11     0.11     0.11 74130181.00   

        rec_count  avg_price  
count    67277.00   67277.00  
unique        NaN        NaN  
top           NaN        NaN  
freq          NaN        NaN  
mean        26.86       0.09  
std         18.22       0.01  
min          1.00       0.07  
25%         13.00       0.08  
50%         22.00       0.09  
75%         36.00       0.10  
max        119.00       0.11

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'length' and 'compression-ratio'.

</div>


In [133]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
df[['Open', 'Close']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [139]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 67277 entries, 0 to 67276
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Ts         67277 non-null  object 
 1   Open       67277 non-null  float64
 2   High       67277 non-null  float64
 3   Low        67277 non-null  float64
 4   Close      67277 non-null  float64
 5   Volume     67277 non-null  float64
 6   Rec_count  67277 non-null  int64  
 7   Avg_price  67277 non-null  float64
dtypes: float64(6), int64(1), object(1)
memory usage: 4.4+ MB


<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                         |
| ----------------- | ------- | ---------- | ---------------------------------------------------------- |
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                                     |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'                       |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe                   |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab                         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
